src/Shared/Components/ComponentDescriptorSerializer.Serialize.cs

```diff
@@ -0,0 +1,121 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Security.Cryptography;
+using System.Text.Json;
+using Microsoft.AspNetCore.DataProtection;
+using Microsoft.AspNetCore.Http;
+
+namespace Microsoft.AspNetCore.Components
+{
+    // ComponentDescriptorSerializer is an abstraction shared by both MVC and Blazor
+    // This file holds the definition used by Mvc to serialize descriptors.
+    internal partial class ComponentDescriptorSerializer
+    {
+        private static readonly object ComponentSequenceKey = new object();
+
+        public ComponentDescriptorSerializer(IDataProtectionProvider dataProtectionProvider) =>
+            DataProtector = dataProtectionProvider.CreateProtector(DataProtectionProviderPurpose);
+
+        public ComponentMarker SerializeInvocation(HttpContext context, Type type, bool prerendered)
+        {
+            var (sequence, descriptor) = CreateSerializedComponentDescriptor(context, type);
+            return prerendered ? ComponentMarker.Prerendered(sequence, descriptor) : ComponentMarker.NonPrerendered(sequence, descriptor);
+        }
+
+        private ComponentDescriptorInvocationSequence GetOrCreateInvocationIdentifier(HttpContext context)
+        {
+            if (!context.Items.TryGetValue(ComponentSequenceKey, out var sequence))
```

Comment:
This is kinda unfortunate that we need to rely on `.Items` for this feature to work correctly no?

Reply:
I can replace it with a scoped service. What is your concern about using .Items?

Reply:
Ok, so I have an alternative for this. I added an `internal IDictionary<object,object> Items` property to ViewContext. The idea is that it is tied to the lifetime of the viewcontext and the children spawn from it and that its meant to be scoped to the page/view/viewcomponent being executed.

I checked all the callsites of the viewcontext ctor to make sure we would keep the same viewcontext for the scope of the page/view/viewcomponent being rendered.

It adds an allocation per page/view/viewcomponent which I think is acceptable. Currently I lazily initialize it, but in practical terms the dictionary will always be allocated as it needs to be shared between parent viewcontext and childrenviewcontext whenever a new childviewcontext gets constructed (which is pretty much anytime you render a partial, layout or similar, so always).

I made the property an IDictionary<object,object> because I didn't want to put something super specific to blazor server-side components into ViewContext. If we wanted to avoid the allocation alltogether we could simply add a property for the sequence generator there as internal, but that feels like bad coupling.

I think the dictionary approach is the best even if we add 1 allocation per view/page/viewcomponent result as is idiomatic (similar to HttpContext.Items) and generally useful even outside of this context (For example for a script manager like primitive). We could consider making it public if we think is generally useful.

For this, I think this approach strikes a balance between practical and decoupled.